File: ts/medium/curry.ts
/*
  Currying is a technique in functional programming where a function with
  multiple arguments is transformed into a sequence of nested functions,
  each taking a single argument.
*/

function main() {
  function curry2(f: Function): Function {
    return function(a: number) {
      return function(b: number) {
        return f(a, b);
      }
    }
  }

  function curry3(f: Function): Function {
    return function(a: number) {
      return function(b: number) {
        return function(c: number) {
          return f(a, b, c);
        }
      }
    }
  }

  function curry(f: Function): Function {
    return function curried(...args: any[]) {
      if (args.length >= f.length) {  //  sum2 (a,b) and c(1,2,3,4,5,6)
        return f.apply(this, args);
      } else {
        return function(...args2: any[]) { //  sum2 (a,b) and c(1)
          const combinedArgs = [...args, ...args2];
          if (combinedArgs.length >= f.length) {
            return f.apply(this, combinedArgs);
          } else {
            return curried(...combinedArgs);
          }
        }
      }
    }
  }

  function sum2(a: number, b: number): number {
    return a + b;
  }

  function sum3(a: number, b: number, c: number): number {
    return a + b + c;
  }

  const carried2 = curry2(sum2);
  const result2 = carried2(1)(2);
  console.log(result2);

  const carried3 = curry3(sum2);
  const result3 = carried3(1)(2)(3);
  console.log(result3);

  const carried33 = curry3(sum3);
  const result33 = carried33(1)(2)(3);
  console.log(result33);

  const carried = curry(sum3);
  const result = carried(1)(9,9);
  console.log(result);
}
main();

File: python/medium/curry.py
""""
Currying is a functional programming technique in which a function is
transformed into a sequence of functions, each taking a single argument.
In Python, currying can be implemented using nested functions or through the
use of decorators and tools from libraries like functools.
"""

def sum_3(a: int,b: int,c: int) -> int:
    return a + b + c

def simple_curry(fn):
    def aFn(a):
        def bFn(b):
            def cFn(c):
                return fn(a,b,c)
            return cFn
        return bFn
    return aFn
curried = simple_curry(sum_3)
result = curried(1)(2)(3)
print(result)

def curry(fn):
    def curried(*args):
        if len(args) == fn.__code__.co_argcount:
            return fn(*args)
        return lambda *args2: curried(*args, *args2)
    return curried

@curry
def sum_3_c(a,b,c):
    return a + b + c
curried_1 = sum_3_c
result_1 = curried_1(1)(2)(3)
print(result_1)
